### PR TITLE
Help: Directly launch apps from their help page + Special help:// URL for better inter-help-page linking

### PR DESCRIPTION
**Pull request by kleinesfilmroellchen**
The main goal of this PR is to improve all kinds of linking from and in between help pages in Help. Some benefits trickle down to man but it's mostly unchanged.

First, the external link handling is improved in Help so that we can introduce direct links to the applications. Second, the new special help:// URL man improve the linking between man pages. This also requires adopting the link translation for the man page website.

# Help: Directly launch an app from its help page
Through links in the help page, the user can directly launch the app whose help page is currently viewed. The idea for this feature came up in the discussion of #11557. The intention is that the user can simply open the app they are currently trying to understand, and play around with it, learn by doing, or follow along with any guide that may be present in the help page. It
feels very great :^)

# Help: Add help://man URLs for links between man pages 
The URLs of the form `help://man/<section>/<page>` link to another help page inside the help application. All previous relative page links are replaced by this new form. This doesn't change any behavior but it looks much nicer :^)

Note that man doesn't handle these new links, but the previous relative links didn't work either.

Re: @linusg @bugreport0 @electrikmilk

### CODE REVIEW

**Comment by AtkinsSJ** [comment]
Very nice! I'm wondering if we need that `associated-apps.ini` file though. It seems like just running the app name as a command should launch it too, but maybe I'm misunderstanding how the PATH works.

**Comment by alimpfard** [comment]
> as that would require some custom Markdown extension

I don't think you need a markdown extension, rather, you can embed them as links with custom protocols (for instance, `help://invoke-program/usr/bin/Eyes`) this is used in Spreadsheet's help already.
For bonus points, you can also make `main` translate these into hyperlinks that invoke LaunchServer :P

**Comment by kleinesfilmroellchen** [comment]
I agree with both of you, associated-apps is not the best solution but the only acceptable one I could think of yesterday

@AtkinsSJ I intentionally didn't want to do the "run app name as command" thing because that's much less flexible. For example, we could instead open the source code for syscalls or open a terminal with `command --help` for command-line applications. Both of those are out of the question with that.

@alimpfard That is a fantastic idea, I'll do that right now! Then this PR will expand to add the ability to `man` as well.

File: Userland/Libraries/LibMarkdown/Text.cpp [hunk 1]
```diff
@@ -156,8 +157,15 @@ void Text::LinkNode::render_for_terminal(StringBuilder& builder) const
 {
     bool is_linked = href.contains("://");
     if (is_linked) {
+        // FIXME: Should/can this live somewhere else?
+        auto url = URL(href);
+        dbgln("{} {}", url, url.path());
```
Comment (alimpfard) [inline Userland/Libraries/LibMarkdown/Text.cpp:162]: That's probably not very useful.

File: Userland/Libraries/LibMarkdown/Text.cpp [hunk 2]
```diff
@@ -156,8 +157,15 @@ void Text::LinkNode::render_for_terminal(StringBuilder& builder) const
...
+        if (url.protocol() == "help" && url.host() == "invoke-program")
+            url = URL::create_with_file_protocol(url.path());
+        dbgln("{}", url.to_string());
```
Comment (alimpfard) [inline Userland/Libraries/LibMarkdown/Text.cpp:165]: same?

File: Base/usr/share/man/man5/SystemServer.md [hunk 3]
```diff
@@ -28,7 +28,7 @@ describing how to launch and manage this service.
...
 * `User` - a name of the user to run the service as. This impacts what UID, GID (and extra GIDs) the service processes have. By default, services are run as root.
 * `WorkingDirectory` - the working directory in which the service is spawned. By default, services are spawned in the root (`"/"`) directory.
-* `SystemModes` - a comma-separated list of system modes in which the service should be enabled. By default, services are only enabled in the "graphical" mode. The current system mode is read from the [kernel command line](../man7/boot_parameters.md#options), and is assumed to be "graphical" if not specified there.
+* `SystemModes` - a comma-separated list of system modes in which the service should be enabled. By default, services are only enabled in the "graphical" mode. The current system mode is read from the [kernel command line](help://man/7/boot_parameters), and is assumed to be "graphical" if not specified there.
```
Comment (alimpfard) [inline Base/usr/share/man/man5/SystemServer.md:31]: The fragment should still be preserved, pls keep that.

**Review by alimpfard — COMMENTED** [review]
LGTM.
Let's wait for CI though~

**Review by alimpfard — APPROVED** [review]

**Review by linusg — COMMENTED** [review]
> As man uses Terminal's built-in URL opener through LaunchServer and LaunchServer doesn't (and shouldn't) understand help:// URLs, the link is automatically rewritten to a normal file:// link for the terminal.

Wait a minute. Don't we get all of this for free in Help via LaunchServer if we use regular `file:///bin/Whatever` URLs in the man pages?! It should forward them to `open_external()`.

**Comment by kleinesfilmroellchen** [comment]
@linusg Well, that was kinda dumb. I have removed the help://invoke-program URLs (through painful rebasing) and adopted Help to properly handle external links especially for these cases.

File: Userland/Applications/Help/main.cpp [hunk 4]
```diff
@@ -144,36 +181,7 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
...
-                tree_view->expand_tree(tree_view_index.value().parent());
-                tree_view->selection().set(tree_view_index.value());
-
-                String page_and_section = manual_model->page_and_section(tree_view_index.value());
-                window->set_title(String::formatted("{} - Help", page_and_section));
-            } else {
-                window->set_title("Help");
-            }
-        });
+        open_url(URL::create_with_url_or_path(path));
```
Comment (linusg) [inline Userland/Applications/Help/main.cpp:184]: If you feel the need to refactor this, please put it in a separate commit.

File: Userland/Applications/Help/main.cpp [hunk 5]
```diff
@@ -221,23 +229,23 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
     };
 
     page_view->on_link_click = [&](auto& url, auto&, unsigned) {
-        if (url.protocol() != "file") {
-            open_external(url);
-            return;
-        }
-        auto path = Core::File::real_path_for(url.path());
-        if (!path.starts_with("/usr/share/man/")) {
+        if (url.protocol() == "file") {
```
Comment (linusg) [inline Userland/Applications/Help/main.cpp:232]: What is actually being changed here? Opening programs with a file URL from Help works on master, I just checked.

Comment (linusg) [reply Userland/Applications/Help/main.cpp:232]: If it's in preparation for the second commit, I don't really see the point of making the change in the first commit.

**Review by linusg — COMMENTED** [review]
Thanks for the update :)

File: Userland/Applications/Help/main.cpp (hunk 5 above)
Comment (kleinesfilmroellchen) [reply Userland/Applications/Help/main.cpp:232]: The main issue with this is that `real_path_for` will cause an unveil violation for `/bin` paths, which I avoid hereby. Relative path support is removed as we don't need it with the new help://man URLs.

Comment (kleinesfilmroellchen) [reply Userland/Applications/Help/main.cpp:232]: But I agree, this should be its own commit.

Comment (linusg) [reply Userland/Applications/Help/main.cpp:232]: > The main issue with this is that real_path_for will cause an unveil violation for /bin paths, which I avoid hereby.

Got it (& should've checked my debug console :sweat_smile:). Legit change then, but please mention the reason in the commit!

**Review by linusg — APPROVED** [review]
Looks great!
